### PR TITLE
reduce number of failed project lookups

### PR DESCRIPTION
**Pull request by dirkmueller**
When a project is assigned that doesn't exist, the projectidcache
was ineffective. we can make that more optimal by first collecting
the set of all projectnames, then looking it up, and then
assigning it to the users.

### CODE REVIEW

File: keystone_json_assignment/json.py [hunk 1]
```diff
@@ -88,25 +88,31 @@ def __init__(self):
 
         self.userprojectmap = {}
         self.useridmap = {}
+
+        projectnames = set()
+        for _, projects in userprojectmap.items():
```
Comment (cmurphy) [inline keystone_json_assignment/json.py]: You could just have `for projects in userprojectmap.values():`

**Review by cmurphy — COMMENTED** [review]
Looks good, could you test it? :)

**Comment by bbobrov** [comment]
How many projects are there in current json file?

**Comment by cmurphy** [comment]
@dirkmueller I fixed the pep8 violation in https://github.com/cmurphy/keystone-json-assignment/pull/4 so you'll have to rebase

**Comment by cmurphy** [comment]
@bbobrov there are 31 projects and 727 users.

**Comment by bbobrov** [comment]
@cmurphy and how many of them do not exist?

I mean, why do we even care to optimize it? :)

**Comment by cmurphy** [comment]
@bbobrov They really should all exist, there's very little reason to have this JSON mapping if the projects don't exist. But it could help in test environments to only have to create a couple of projects instead of all 31. @dirkmueller did you run into a problem like this on the production env?

**Comment by dirkmueller** [comment]
@cmurphy no, its just something I hit on the tst env where I didn't create the projects and it spends about 30s on hanging, which is not so nice. 

I'll test it.

**Comment by dirkmueller** [comment]
Its tested now :)

**Review by cmurphy — APPROVED** [review]

**Comment by cmurphy** [comment]
Needs a rebase

**Comment by dirkmueller** [comment]
@cmurphy updated

**Pull request merged**